Rewrote do_sshfs_mount() to reuse the existing SSH session

JeffreyMitchellSL opened the pull request:
Resolves #95, #82, and #31.

Reworked `do_sshfs_mount()` to reuse a single SSH connection for SSHFS instead of creating a second connection. This makes the process more reliable and efficient while reducing dependencies. Because there's no guest-to-host connection, the host no longer requires a running SSH server, only the `sftp-server` executable, which should allow shared folders inside docker containers. On my machine, these changes bring down the time to launch and immediately shutdown a CentOS VM with 20 shared folders from 20 seconds to 19.

Enabled SSHFS tests in CI.

Disabled ssh-agent tunneling for SSHFS mounts.

I used stderr markers before launching sshfs to distinguish between SSH errors and SSHFS errors.

ALSchwalm commented on transient/sshfs.py:
```diff
@@ -115,127 +108,92 @@ def provision_system(timeout: int, ssh_config: ssh.SshConfig, is_slow: bool) ->
 
 
+def get_sftp_server() -> str:
+    for path in [
```
I think we should do this the same way we do ssh itself. Just have a default of 'sftp-server' (i.e., just assuming it's in the path) and allow the user to specify otherwise. Definitely don't just have a hardcoded list of absolute paths to check.

JeffreyMitchellSL replied:
Sounds good to me. I made the relevant change.

ALSchwalm commented on transient/linux.py:
```diff
@@ -22,5 +22,5 @@ def prctl(option: int, arg2: int = 0, arg3: int = 0, arg4: int = 0, arg5: int =
 
 def set_death_signal(signal: int) -> int:
-    """Send `signal` to this process when the parent dies"""
+    """Send `signal` to this process when the parent thread dies"""
```
To be clear, the child will receive the signal when the parent processes dies. That process may be a 'thread', but that doesn't really matter.

ALSchwalm commented on transient/sshfs.py:
```diff
@@ -115,127 +108,92 @@ def provision_system(timeout: int, ssh_config: ssh.SshConfig, is_slow: bool) ->
+    client = ssh.SshClient(sshfs_config, command=ssh_command)
+
+    logging.info(f"Sending sshfs mount command '{sshfs_command}'")
+    ssh_proc = client.connect(
```
The trouble with not running the prctl call here (and the reason we have that in general) is that one of the main goal of transient is to avoid leaving state behind when it's done. This includes zombie processes. From my reading of this, if the transient process is killed after this call, the ssh subprocess may keep running. We need to ensure this doesn't happen.

JeffreyMitchellSL replied:
I think it's unlikely that this SSH session would persist indefinitely. Still, I agree that we should avoid leaving any zombie processes at all. My new approach allows the prctl call by leaving the SSHFS threads running until Transient exits.

ALSchwalm commented on transient/sshfs.py:
```diff
@@ -115,127 +108,92 @@ def provision_system(timeout: int, ssh_config: ssh.SshConfig, is_slow: bool) ->
+    # Everything from here on out simply verifies that nothing went wrong.
+    if ssh_proc.stderr is not None:
+        first_stderr_line = ssh_proc.stderr.readline().decode("utf-8")
+        if "TRANSIENT_SSH_COMPLETE" not in first_stderr_line:
```
Just assuming the first line of stderr will be this string seems somewhat questionable. It seems like SSH itself could write to stderr. However, I'm not sure it is worth a bunch of complexity to avoid that if we've never actually seen it happen. Still something to think about.

JeffreyMitchellSL replied:
True. I haven't encountered any such SSH-based stderr that doesn't prevent a connection, and Transient seems to do a good job of proactively passing options to suppress it.

ALSchwalm commented on transient/sshfs.py:
```diff
@@ -115,127 +108,92 @@ def provision_system(timeout: int, ssh_config: ssh.SshConfig, is_slow: bool) ->
-        stderr = raw_stderr.decode("utf-8")
+    # SSHFS command failed
+    if ssh_proc.returncode:
+        print("SSHFS returned {0}".format(ssh_proc.returncode))
```
Given that this is part of the exception, it shouldn't be printed here (and you shouldn't use print in general. Use logging.error for this sort of thing).

JeffreyMitchellSL replied:
Deleted.

ALSchwalm commented on transient/sshfs.py:
```diff
@@ -115,127 +108,92 @@ def provision_system(timeout: int, ssh_config: ssh.SshConfig, is_slow: bool) ->
 
-        # Check whether this failure looks like it is the result of the vm being
-        # unprovisioned
+        # Check stderr
```
It seems like this was more clear before

JeffreyMitchellSL replied:
Reverted.

ALSchwalm commented on transient/ssh.py:
```diff
@@ -138,6 +139,10 @@ def __timed_connection(
+                preexec_fn = None
+                if tie_to_thread:
+                    preexec_fn = lambda: linux.set_death_signal(signal.SIGTERM)
+
```
I'm very concerned about this. I don't think there is probably ever a situation where we should not set the death signal for an ssh subprocess, as the transient process could die at any point (in the worst case, because a user sends SIGKILL), and we should still avoid leaving processes running.

JeffreyMitchellSL replied:
I managed to remove this by leaving SSHFS threads dormant beyond the mount setup so that `set_death_signal` wouldn't kill them.

ALSchwalm commented on transient/sshfs.py:
```diff
@@ -1,15 +1,15 @@
-_PROVISION_TIME = 120
-_PROVISION_TIME_SLOW = 300
+SSHFS_MAX_RUN_TIME = 0.5
+PROVISION_TIME = 120
```
These should probably not be public.

JeffreyMitchellSL replied:
Reverted.

ALSchwalm commented on transient/sshfs.py:
```diff
@@ -115,127 +108,92 @@ def provision_system(timeout: int, ssh_config: ssh.SshConfig, is_slow: bool) ->
-        conn.poll()
+    # Everything from here on out simply verifies that nothing went wrong.
+    if ssh_proc.stderr is not None:
+        first_stderr_line = ssh_proc.stderr.readline().decode("utf-8")
```
I believe this will block forever if the guest hangs (because of an sshfs problem causing it to hang, for example).

ALSchwalm replied:
Actually, I think it would be much better to have this be a timed blocking read (via `communicate`). This should also eliminate my concern below about something else writing to stderr. What would be ideal here would be a line buffered read with a timeout. That way we could set the timeout be be high, because it would almost always return almost immediately when the guest writes the magic line. However, I'm not sure it's possible to do that without more hacks than I'm willing to have.

JeffreyMitchellSL replied:
This should not hang. SSH will either timeout or get connection refused, and SSHFS is not yet called at this point.
This also acts as a dividing line between waiting for SSH, which takes a variable amount of time and will timeout on its own based on the user's arguments, and waiting for SSHFS, which is very fast and needs its own timeout. Using `communicate()` would always block for the maximum time unless something breaks, which would slow things down.

ALSchwalm commented on transient/sshfs.py:
```diff
@@ -115,127 +108,92 @@ def provision_system(timeout: int, ssh_config: ssh.SshConfig, is_slow: bool) ->
+    try:
+        ssh_proc.wait(SSHFS_MAX_RUN_TIME)
+    except subprocess.TimeoutExpired as e:
+        # This is expected.
```
This should be a more descriptive comment about _why_ this is an expected failure. Something like:

```
Because SSHFS is communicating over stdin of the ssh connection, SSH needs to run for
as long as the mount exists. Therefore, we can't wait until connection is closed. Instead 
we wait a short time so the script has a chance to output a sentinel value. So the timeout 
here is expected.
```

JeffreyMitchellSL replied:
I agree. I have improved this.

ALSchwalm commented on transient/sshfs.py:
```diff
@@ -115,127 +108,92 @@ def provision_system(timeout: int, ssh_config: ssh.SshConfig, is_slow: bool) ->
 
-        stderr = raw_stderr.decode("utf-8")
+    # SSHFS command failed
+    if ssh_proc.returncode:
```
What if ssh_proc exited with 0? Doesn't the fact that ssh exited _at all_ indicate something is wrong? We would expect it to continue until we kill it, right?

JeffreyMitchellSL replied:
Hmm, true. At first I thought that graceful exits should be ok, but you're right. The user expects an SSHFS mount, and exiting would prevent that. I fixed this in my recent commits.

ALSchwalm commented on transient/sshfs.py:
```diff
@@ -115,127 +108,92 @@ def provision_system(timeout: int, ssh_config: ssh.SshConfig, is_slow: bool) ->
-                    ssh_config=ssh_config,
-                    is_slow=True,
-                )
+    # We already waited for the timeout
```
I'd like a little more description from this. Maybe "Sanity check that the ssh connection has not closed unexpectedly before we return"

JeffreyMitchellSL replied:
I agree. I have improved this.

ALSchwalm commented on transient/sshfs.py:
```diff
@@ -115,127 +108,92 @@ def provision_system(timeout: int, ssh_config: ssh.SshConfig, is_slow: bool) ->
-        # Ensure returncode is set
-        conn.poll()
+    # Everything from here on out simply verifies that nothing went wrong.
+    if ssh_proc.stderr is not None:
```
Why would stderr ever be None? You setup the subprocess with PIPE

JeffreyMitchellSL replied:
It wouldn't, and this bugs me too. I had to add this if statement to appease mypy.

ALSchwalm commented on transient/transient.py:
```diff
@@ -374,12 +373,21 @@ def run(self) -> None:
             absolute_local_path = os.path.abspath(local)
+
+            # Our preexec_fn kills the server early if it's launched from the thread.
+            sftp_proc = subprocess.Popen(
```
Given that this is an implementation detail of the folder mounting, I'm not sure it should be spawned here.

JeffreyMitchellSL replied:
Fixed by leaving SSHFS threads running.

ALSchwalm reviewed (changes requested):
This is looking good. It'll be cool once it's merged. I do have a few changed and concerns though.

Also, can you update this to make the shared folder test run in the docker container?

JeffreyMitchellSL commented:
> This is looking good. It'll be cool once it's merged. I do have a few changed and concerns though.
> 
> Also, can you update this to make the shared folder test run in the docker container?

Due to the fact that the preexec function kills subprocesses when the parent _thread_ dies (See the PDEATHSIG warning in https://man7.org/linux/man-pages/man2/prctl.2.html), I changed my approach to how SSHFS threads work. Now, SSHFS threads persist in a dormant state, relaying mount completion back to the main thread through Python's `Event` class instead of `wait()`. `checked_threading` was too generic for this purpose.
In order to clean up `transient.py`, I also switched to using a semaphore to prevent overloading the SSH server. By being conservative about when exactly I lock the semaphore, I further sped up the time to launch/kill my VM with 20 shared folders to 19.9 seconds.
My most recent changes and responses should resolve all of your comments.

~~Something about these changes made it more likely that #101 gets hit, so I also included my fix for that issue in a separate commit.~~

ALSchwalm commented on transient/sshfs.py:
```diff
@@ -115,127 +118,131 @@ def provision_system(timeout: int, ssh_config: ssh.SshConfig, is_slow: bool) ->
-    conn = client.connect_piped(timeout=connect_timeout)
+def get_sftp_server(name: str) -> str:
+    # sftp-server isn't on the default PATH. This adds the places it gets installed.
+    path_addon = ":".join(["/usr/lib/openssh", "/usr/libexec/openssh", "/usr/libexec"])
```
Hmm, the trouble with this is that it doesn't work for systems that don't use one of these spots. For example, archlinux puts it in "/usr/lib/ssh". In general, this is pulled from the `Subsystem	sftp` bit of sshd_config. I'm not sure it's really necessary that we read that to find the correct spot (as sshd_config is not required to be world readable anyway, so that introduces new problems).  Ultimately I'd like "/usr/lib/ssh" added to this list and for the error below to suggest using the `sftp-bin-name` flag.

JeffreyMitchellSL replied:
Done.

ALSchwalm commented on transient/sshfs.py:
```diff
@@ -21,6 +26,11 @@
 
+global provision_lock
+provision_lock = threading.Lock()
+global provision_success
```
To my knowledge, this doesn't do anything? A quick test indicates that:

```
>>> foo = 10
>>> global foo
>>> 
>>> 
>>> def func():
...   foo = 100
... 
>>> func()
>>> foo
10
```
The `global` key word must still be used in the context where it would otherwise be shadowed.

JeffreyMitchellSL replied:
That is correct and it bugs me too. However, like `if ssh_proc.stderr is not None:`, mypy, and therefore `make check`, fail if I don't include those global lines even if they contribute nothing to the program.
Luckily, putting these variables into SshfsThread, like you suggested in another comment, allows us to avoid the issue.

ALSchwalm commented on transient/sshfs.py:
```diff
@@ -115,127 +118,131 @@ def provision_system(timeout: int, ssh_config: ssh.SshConfig, is_slow: bool) ->
-                    ssh_config=ssh_config,
-                    is_slow=True,
+            # Everything from here on out simply verifies that nothing went wrong.
+            if ssh_proc.stderr is not None:
```
Again, I don't see how this can be None. You have set the stderr flag to PIPE, so it will never be None (and I'm not sure why we would skip this if it were None).

JeffreyMitchellSL replied:
It can't be None, and this bugs me too. However, mypy doesn't look any further than the stated return type of connect(), which superficially says it might have a null stderr. Because mypy can't figure out that stderr will never be None, `make check` fails without this.

I did, however, figure out that changing from an if to an assert allows me to reduce the indentation of the lines in that block without provoking mypy.

ALSchwalm commented on transient/sshfs.py:
```diff
@@ -115,127 +118,131 @@ def provision_system(timeout: int, ssh_config: ssh.SshConfig, is_slow: bool) ->
+            if ssh_proc.stderr is not None:
+                first_stderr_line = ssh_proc.stderr.readline().decode("utf-8")
+                if "TRANSIENT_SSH_COMPLETE" not in first_stderr_line:
+                    stderr = first_stderr_line + ssh_proc.communicate()[1].decode("utf-8")
```
This communicate call can hang forever (which seems plausible if for some reason we are getting a line that isn't the expected one)

JeffreyMitchellSL replied:
Fixed.

ALSchwalm commented on transient/sshfs.py:
```diff
@@ -115,127 +118,131 @@ def provision_system(timeout: int, ssh_config: ssh.SshConfig, is_slow: bool) ->
+
+        try:
+            # Now that the SSH connection is established, the SSHFS timeout can be very short
+            stderr = ssh_proc.communicate(timeout=_SSHFS_MAX_RUN_TIME)[1].decode("utf-8")
```
Wouldn't it be better to add another line of output to indicate that sshfs has started (after the call to sshfs)? Then we can do the same as above rather than expecting a timeout.

JeffreyMitchellSL replied:
That would be nice, but SSHFS needs to stay in the foreground. Adding an `&` to push sshfs into the background would not affect speed because the marker would not accurately indicate a successful mount. It would also disrupt regular error handling.

After I figured out that the CI SSHFS tests were failing because VMs without KVM are attrociously slow, I did add another marker, though it's directly before, not after, starting SSHFS so that the timeout can be much shorter than a non-KVM VM would normally require.

ALSchwalm commented on transient/transient.py:
```diff
@@ -379,20 +372,15 @@ def run(self) -> None:
 
         for sshfs_thread in sshfs_threads:
-            sshfs_thread.join()
+            sshfs_thread.is_complete.wait()
```
I'd rather have a wait method on the thread that waits on is_complete and does the lines below

JeffreyMitchellSL replied:
Done.

ALSchwalm commented on transient/sshfs.py:
```diff
@@ -21,6 +26,11 @@
+global provision_lock
+provision_lock = threading.Lock()
+global provision_success
+provision_success = None
```
This should be type annotated. In general I'm ok with things like `provision_lock = threading.Lock()` not being annotated, as the type is right there, but for things like `None` we know the value is definitely not always going to be `None`, so knowing the actual type is useful.

ALSchwalm replied:
Also, this and the above provision_lock seems like they should be class variables on `SshfsThread` given that the semaphore is there already anyway

JeffreyMitchellSL replied:
Done and done. Luckily, this also allows us to avoid mypy's insistence on `global` keywords here.

ALSchwalm commented on transient/sshfs.py:
```diff
@@ -115,127 +118,131 @@ def provision_system(timeout: int, ssh_config: ssh.SshConfig, is_slow: bool) ->
+        # Wake up the main thread
+        self.is_complete.set()
+
+        if ssh_proc.wait():
```
Hmm, maybe we want to log the error here as well, as the actual exception may never been seen, right?

JeffreyMitchellSL replied:
No. Whenever a Python thread hits an unhandled exception, the output goes to stderr, regardless of whether it's the main thread or a daemon thread. Any time this hits an SSHFS error, it always prints it at the end of a Python traceback.

ALSchwalm reviewed (changes requested):
A few more comments. I like the `SshfsThread` concept with a semaphore more than the previous approach, nice work!

That said, I'm beginning to question whether the multi thread sshfs mount stuff is really worth it. I definitely want to get the logic to re-use the existing ssh connection in, but I have concerns about the complexity that has when interacting with the multithreaded setup. Especially given that the sshfs mount is something that it is actually plausible could fail (and that's not even getting in to the provisioning stuff).

JeffreyMitchellSL commented:
SSHFS tests were failing in CI on this branch because, in CI, the VMs run without KVM, making them incredibly slow. As such, the guest didn't even finish searching for `sudo` and `sshfs` on the path before the SSHFS timeout expired. I fixed this by adding a call to `which sudo sshfs` to cache those paths before emitting another stderr marker and starting sshfs. I still had to slightly increase the SSHFS timeout, which slows things down a little.
My most recent changes and responses should resolve all of your comments.

It looks like this PR should also close #31, because it enables SSHFS tests in CI, and #82, because it significantly improves error clarity for SSHFS failures and resolves both specific examples listed in the issue.

JeffreyMitchellSL committed:
Created SshfsThread to reuse existing SSH connections instead of creating new ones
    Removed the need for a running SSH server on the host
    Removed is_slow because the timeout that sets it to true can no longer happen
    Disabled ssh-agent tunneling for SSHFS mounts
    Removed local_user and local_password because the guest no longer needs access to the host
    Updated shared-folders.md to reflect the new SSHFS method and reduced restrictions
    Deleted docker.md because there are no more special cases to prevent Transient in containers
Enabled SSHFS tests in CI
    Extended CI SSH timeout to accomodate the extra SSH traffic
Added option to specify the sftp server binary's name
Switched to semaphore-based enforcement of MAX_CONCURRENT_SSHFS
    Moved MAX_CONCURRENT_SSHFS to sshfs.py
Clarified a comment in linux.set_death_signal() to indicate that the parent THREAD is what matters

JeffreyMitchellSL committed:
Further increased CI SSH timeout to 780 seconds

ALSchwalm commented:
Thanks for working on this! I'm sorry it took so long to land, but this has finally been merged as of d9dc0915723f066ba77742329fc9b77915e43327
